### ISSUE TITLE
performance degradation during use

### ISSUE THREAD

**Issue by maxalone** [Luidog/node-red-contrib-filemaker#22]
just after the first fast actions, over the hours the performance clearly deteriorates.
it goes from 0 seconds to 10 seconds or more at the end of the day for each FIND EDIT DELETE ADD etc. making it useless for production. imagine what happens when there are two linked actions in a Flow.
the only system I found to re-enable fast access is to restart Red.
the value of Concurrency in Data-Api-Configuration seems to have no influence on this behavior.
I also ask if someone has already tested with FMS 18.0.2 (I haven't updated since 18.0.1.123)
thank you for your precious work

**Comment by Luidog** [comment]
@maxalone, Thank you for filing an issue for this. I have the current version of these nodes working in a few places and I am not able to replicate this issue.

So that we can triage the issue can you please provide a test flow that showcases the issue? 

It would also be good for you to provide the specifications for your FileMaker server and update it to the latest version of server (18.0.2.217).

**Comment by maxalone** [comment]
Thanks for the attention

each flow slows down the overall performance during the day.
simple flows need to be invoked several times (like the one below with a simple extraction for a list) and the slowdown becomes unacceptable after a few hours of executions
![image](https://user-images.githubusercontent.com/28682199/62892017-256cfe00-bd3f-11e9-92a7-e0a65dd4e9bd.png)

from
![image](https://user-images.githubusercontent.com/28682199/62892632-a678c500-bd40-11e9-925c-b24e40cb9d90.png)
to
![image](https://user-images.githubusercontent.com/28682199/62892094-52211580-bd3f-11e9-8886-6c76a636b45f.png)

others, like this image parser, increase the delay even after a few first executions

![image](https://user-images.githubusercontent.com/28682199/62892141-69f89980-bd3f-11e9-99a7-2ea003b1baad.png)

from 
![image](https://user-images.githubusercontent.com/28682199/62892589-8f39d780-bd40-11e9-9226-7550f8041e1f.png)
to
![image](https://user-images.githubusercontent.com/28682199/62892457-39652f80-bd40-11e9-804a-b1676a12243b.png)

looks like a resource release problem but I failed to notice any increasing use of RAM or CPU either by FMS or by Node or Red
initially I thought it was strictly related to the number of active API + PRO users

as I said we have FMS currently in version 18.0.1.123
if you need further configuration information please ask me

thanks again

**Comment by Luidog** [comment]
@maxalone thank you for the detailed information. It might take me a few days to replicate this issue. 

Since I first commented on the issue I have been running a simple flow that creates a record every two seconds.  After about 4 hours I am seeing inconsistent results. It looks like there is a delay that occurs, though I have not yet identified why. It is interesting that the delay does eventually correct itself.

 Would you let me know what version of Node-RED you are using?

**Comment by Luidog** [comment]
@maxalone  I have can't say conclusively what is causing the slowdown but it does look to be related to a slowdown in FileMaker Data API responses. 

Using the flow I mentioned earlier I am able to identify a point where the Data API's responses begin to be delayed.

When you get a chance can you expand on the scope of the find you are performing? I am specifically looking for the following information:

Are the fields you are using in the find indexed?
Can you provide an example of a typical find request?
How many records are in the table where you are performing the find?
Does the table contain any related information?
How often are you performing this find?

My current theory is that the token used in each request contains a history of all previous requests, once the history gets large enough it slows down the Data API responses. If I can conclusively identify the cause of the response slowdown I can make adjustments to the underlying client.

**Comment by maxalone** [comment]
I'll give you all the detailed information shortly, in these hours I'm traveling and I can't.  thanks for your interest

**Comment by maxalone** [comment]
the answers:

**Are the fields you are using in the find indexed?** yes

**Can you provide an example of a typical find request?**  "query":[{"chiaveatt": uniquekey}]
**How many records are in the current found set?** 1
there are few searches that produce lists of records

**How many records are in the table where you are performing the find?** 904 (the biggest)
**Does the table contain any related information?** if you mean portals, yes

**How often are you performing this find?** it depends on how many connected users (for now it is only in development) and on the requested function, ranging from a request per session (expected one session / time in production) up to 25/30 uses of the same flow per request (like the image parser in example)
  when developing I open a new session about every 5 minutes on average

**Comment by Luidog** [comment]
Thank you. I am traveling for the next few days, but I will continue to triage when I get back. 

I have confirmed a slowdown of responses over time with create requests. I will use the answers to my questions to replicate the slowdown with find requests as well. 

I will also test with the latest version of FileMaker Server 17 to see if the issue originated in FileMaker Server 18.

**Comment by Luidog** [comment]
My tests did confirm an issue which would cause the concurrency configuration to not be respected. In the tests where the concurrency configuration was not respected the default concurrency of 1 would be set.

With a concurrency of 1, it is very easy to see performance degradation. The FileMaker Data API responses will vary and become increasingly delayed overtime until a threshold is exceeded. Once the threshold is exceeded the response times decrease to their original levels. Depending on the rate of requests for client node a concurrency of 1 could cause substantial performance degradation to a flow.

@maxalone would you mind updating to node-red-contrib-filemaker 2.1.0 and configuring the client with a larger concurrency setting and commenting here with results? Response times may increase as new sessions are made available but you shouldn't see overall performance degradation until each session is in use.

**Comment by Luidog** [comment]
I have a find request test running that is similar to the example @maxalone provided. The test uses a client configured with a timeout of 1000 and a concurrency of 25. The test issues 25 random find requests to a set of 9137 records. 

The two test failure criteria are a find request taking longer than the client timeout configuration or the overall set of 25 finds taking longer than 26 seconds.

**Comment by maxalone** [comment]
here I am, I proceeded to update all the libraries and restart red and fms

I used the image parser because it is what stresses most, in an "intensive" way (8 sessions in 10 minutes, each session unleashes 21 calls to the parser stream extracting containers from a record after a find on 940 records) and in the end the delay is similar to the previous times
I used 20 in concurrency and no timeout
I repeated the test twice, after a reboot, and the results were similar

![image](https://user-images.githubusercontent.com/28682199/63215003-0217c980-c118-11e9-9701-d7f217aad148.png)

I then set concurrency to 2 and repeated the test (8 sessions of 21 parsers each, in 10 minutes)
identical result except for a greater possibility of seeing the flow interrupted due to API unavailability

![image](https://user-images.githubusercontent.com/28682199/63215255-d0086680-c11b-11e9-9cf7-eca9f001f904.png)

without entering the cabal, the maximum delay at each execution increased by 7 seconds

I trust a solution

**Comment by maxalone** [comment]
I tried to set a timeout of 1000 and repeat the tests with negligible differences in the final results

a curious thing I want to point out is the number of DATA API connections that result in the FMS dashboard
if imposed 20 in concurrency, groups of 20 connections are created proportionately (I also saw 96 connections, and we have 5 licenses). these connections are reset only if I restart RED, otherwise it seems to me to fall automatically after about 15 minutes.

in an incomprehensible way after some RED restarts only one DATA API connection is shown (or at most, three) even if I set it to 40 in concurrency

**Comment by Luidog** [comment]
I have been running 25 random find requests every two minutes for the past 5 hours. I am not seeing any performance degradation. If the tests reach a full 8 hours without any failures I am going to close this issue.

@maxalone Thank you for sharing your results. Based on what was shared it looks like the issue is either environmental (specific to your FileMaker or Node-RED server) or specific to your flow.  The information I have causes me to think the issue could be related to either the extract container node or another node that is not part of this project. If the extract container node is using a buffer object try deleting that buffer object after each use.

If you can provide a test flow and FileMaker file that conclusively causes a degradation in performance over time we can open and revisit this issue. 

I appreciate you taking the time to report this issue and continuing to work with me to help identify it. In the course of working on this issue, I have identified a few upgrades which will make this project even better:

As the project currently uses an in-memory data store provided by [marpat] (https://github.com/Luidog/marpat) here https://github.com/Luidog/node-red-contrib-filemaker/blob/aba8ad7968fe6685f8469af6ee4d71f107786485/src/client/client.js#L4 every restart of Node-RED clears the configured clients.  I have identified changes that need to be made to enable this project to use any data store allowed by [marpat] (https://github.com/Luidog/marpat).

Each restart effectively causes clients to open a new session to the configured concurrency. FileMaker Server's reaction to this has been *_interesting_*. I look forward to learning more. Tracking and maintaining active sessions is critical to server performance.  The current behavior used in this project is currently the safest approach. 

New sessions are only created when there are sessions queued and the current number of sessions is less than the concurrency setting. This provides a balance between response times and server usage. The test I am currently running has been capable of sending 25 requests in under 26 seconds every two minutes for the last 5 hours. It started at 25 connections but is currently running on only one. If the [watch](https://github.com/Luidog/fms-api-client/blob/8b8d42cc1253f437f4e709a14dca8a331028a8ce/src/models/agent.model.js#L400-L453) method detects pending requests and there are no sessions available it will start a new session.

Thankfully Claris also provided an Admin API  that can administer client sessions. Work for a session maintainer will be done in [fms-admin-client](https://www.github.com/luidog/fms-admin-client). Once that is complete I will use that project alongside [fms-api-client](https://www.github.com/luidog/fms-api-client) to maintain active sessions via a specific account. This will provide greater control over session usage and server performance.

**Comment by maxalone** [comment]
meanwhile, thanks for all the research work done

I just want to point out that, as I wrote in the first report, the delay is also accumulated using only find and extracting text lists, but to get a measurable delay I have to wait a few hours of work
 the image parser flow shortens the time but the problem also exists with other extraction flows.

as far as the flow buffer is concerned, it is obviously not possible to delete it because it is the only object that must be returned and it is the purpose of the flow itself.
the delay has nothing to do with the process of production (clearly visible during the tracing of the executions) of the buffer but proportionally in the invocation of the various modules that communicate with FMS.

I am sure that the direction you have taken is the right one because you are focusing on managing connection sessions, which are reset when RED is restarted, and in balancing between active and to be created sessions.

I will keep you updated on every useful element.
thanks again for everything

**Comment by Luidog** [comment]
@maxalone If you can make me a test flow that replicates the issue I can continue to triage the issue. 

Again it sounds like the performance degradation is either environmental or specific to your individual flow. Try using the extract container node to write the file to the filesystem instead of to a buffer object.  Creating a buffer object without being destroyed at some point can have adverse side effects like performance degradation. The Node-RED's runtime's garbage collector will eventually clean up old buffers but you will see increasing performance losses.

Perhaps you might have missed it in my long comment above, but I am working on persisting the clients between Node-RED restarts. This functionality was originally slated for 2.0.0 but needed to be put on hold to address a scalability issue with the Data API. As you note, once clients are persisted connection management will make more sense.

**Comment by Luidog** [comment]
I have been performing 25 random find requests every two minutes for the past 25 hours. Each set of 25 finds completed in 20-24 seconds. No find request set took longer than 26 seconds. This test can be found here: https://github.com/Luidog/node-red-contrib-filemaker/blob/master/examples/flows/find-performance-test.json

I also performed a create request every 2 seconds for twelve hours.  This test can be found here: https://github.com/Luidog/node-red-contrib-filemaker/blob/master/examples/flows/find-performance-test.json

Using the 2.1.0 release I can not replicate a performance degradation over time. I am closing this issue.

**Comment by Luidog** [comment]
@maxalone I know you are still experiencing an issue with performance degradation. I will continue to triage that issue as I have time.

Please review my previous comments on environmental and process issues that might be causing the performance degradation. I think you might be able to avoid degradation by revising your process a bit. Please try saving the extract files to the filesystem instead of to a buffer object.

If the issue persists you might look at using the underlying [fms-api-client](https://github.com/luidog/fms-api-client) directly. This will give you much more fine-grain control over session persistence, file extraction, and the overall process.

I ask that you please provide a test flow similar that can independently replicate the issue.  If you are willing to do so please open up a new issue here specific to the extract container node and attach the flow and a link to the corresponding filemaker file for testing.
